### PR TITLE
Feature/vmap training

### PR DESCRIPTION
**Pull request by Drenderer**
# Purpose
This PR implements an option to vmap the `make_step` function in the klax training loop over the `TrainingState`. This allows training an ensemble (multiple instances of the same model) simultaneously, in a single call to `klax.fit`.

# Changes
- The `run_training_loop` function now accepts a `step_function` argument, which is used to update the `TrainingStep`. This enables the use of custom update step logic, and to apply function transformations such as jit or vmap to the `step_function` before passing it to the training loop. Internally, `klax.fit` simply passes `step_function=make_step` to `run_training_loop`.
- `klax.fit` is extended with two keyword arguments: `jit_compile` and `vmap_ensemble` which are `True` and `False` by default, respectively. `jit_compile=True` jit-compiles the step_function passed to `run_training_loop`. `vmap_ensemble=True` vmaps the `step_function` over the `TrainingState`, `optimizer.init` over the model parameters, and the default loss and validation loss metric evaluation functions over the model parameters. All together, this ensures that a training run with `vmap_ensemble` cleanly vectorizes the training over an ensemble. (Note that the same batch is used for all models in the ensemble in each step, i.e., the batcher is not vmaped).
- To cleanly vmap over the `TrainingState` the `step` property was moved from `TrainingState` to `TrainingContext`. This has multiple positive knock-on effects:
    - `make_step` is no longer responsible for incrementing the step. Instead, this is done in `run_training_loop` - a much more natural choice.
    - `step` is no longer required to be a jax array. This gives a much cleaner history, which now contains a list of integers as steps, instead of the previous list of scalar jax arrays.
    - Previously, accessing the current step in a Callback was done via `contest.state.step` and accessing the total number of planned training steps was `context.steps`. Now we have `context.step` and `context.steps`.
    - However, this introduces a breaking change for all existing callbacks that use the step count.

# Todo
- [x] Add documentation
- [x] Add tests

### CODE REVIEW

**Comment by jaosch** [comment]
The PR got contaminated with changes from #106 . This should be reverted before merging.

**Comment by Drenderer** [comment]
> The PR got contaminated with changes from #106 . This should be reverted before merging.

Fixed, will merge now.
